daphne/src/lib.rs

```diff
@@ -429,16 +444,114 @@ pub struct DapTaskConfig {
+            vdaf_config: messages::taskprov::VdafConfig {
+                dp_config: messages::taskprov::DpConfig::None,
+                var: messages::taskprov::VdafTypeVar::Prio2 { dimension: 10 },
+            },
+        };
+
+        let encoded_taskprov_config = taskprov_config.get_encoded_with_param(&self.version);
+        let task_id = taskprov::compute_task_id(self.version, &encoded_taskprov_config);
+
+        // Compute the DAP task config.
+        let task_config = DapTaskConfig::try_from_taskprov(
+            self.version,
+            &task_id,
+            taskprov_config,
+            vdaf_verify_key_init,
+            collector_hpke_config,
+        )
+        .unwrap();
+
+        let (taskprov_advertisement, taskprov_report_extension_payload) = match self.version {
+            DapVersion::Draft07 => (Some(encode_base64url(&encoded_taskprov_config)), Vec::new()),
+            // draft02 compatibility: The taskprov config is advertised in an HTTP header in
+            // the latest draft. In draft02, it is carried by a report extension.
+            DapVersion::Draft02 => (None, encoded_taskprov_config),
+        };
+
+        Ok((
+            task_config,
+            task_id,
+            taskprov_advertisement,
+            taskprov_report_extension_payload,
+        ))
+    }
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+impl Default for DapTaskParameters {
+    fn default() -> Self {
+        Self {
+            version: Default::default(),
+            leader_url: "https://leader.example.com/".parse().unwrap(),
+            helper_url: "https://helper.example.com/".parse().unwrap(),
+            time_precision: 3600, // 1 hour
+            lifetime: 86400 * 14, // two weeks
+            min_batch_size: 10,
+            query: DapQueryConfig::TimeInterval,
+            vdaf: VdafConfig::Prio2 { dimension: 10 },
+        }
+    }
+}
+
+/// Per-task DAP parameters.
+#[derive(Clone, Deserialize, Serialize)]
+#[cfg_attr(test, derive(PartialEq, Debug))]
+pub struct DapTaskConfig {
+    /// Same as [`DapTaskParameters`].
+    pub version: DapVersion,
+    pub leader_url: Url,
+    pub helper_url: Url,
+    pub time_precision: Duration,
+    pub min_batch_size: u64,
+    pub query: DapQueryConfig,
+    pub vdaf: VdafConfig,
+
+    /// The time at which the task expires.
+    pub expiration: Time,
 
     /// VDAF verification key shared by the Aggregators. Used to aggregate reports.
     pub vdaf_verify_key: VdafVerifyKey,
 
     /// The Collector's HPKE configuration for this task.
     pub collector_hpke_config: HpkeConfig,
 
-    /// If true, then the taskprov extension was used to configure this task.
+    /// Deprecated. Indicates that the task was configured via draft-wang-ppm-taskprov.
+    #[serde(default, rename = "taskprov")]
+    pub deprecated_taskprov: bool,
+
+    /// Method by which the task was configured.
     #[serde(default)]
-    pub taskprov: bool,
+    pub method: DapTaskConfigMethod,
 }
```

Comment:
It would be nice if these two fields were somehow coalesced. There seems to be quite a fair number of invalid states that are possible between these fields and `version` and they are all `pub`. Could we encapsulate this a bit more so this type is more obvious in how it's supposed to be used?

Reply:
Which invalid states?

Note that one requirement I'm trying to fullfill with this change is maintaining backwards compatibility with `DapTaskConfig`s that are already in KV. This means the old JSON object has a field called `taskprov` and the new one has a field called `method`. This change has code that uses the old field if available.

Reply:
See [this comment](https://github.com/cloudflare/daphne/pull/436#discussion_r1406308160)